**Table of contents**<a id='toc0_'></a>    
- 1. [Problem 1: Production economy and CO2 taxation](#toc1_)    
- 2. [Problem 2: Career choice model](#toc2_)    
- 3. [Problem 3: Barycentric interpolation](#toc3_)    

<!-- vscode-jupyter-toc-config
	numbering=true
	anchor=true
	flat=false
	minLevel=2
	maxLevel=6
	/vscode-jupyter-toc-config -->
<!-- THIS CELL WILL BE REPLACED ON TOC UPDATE. DO NOT WRITE YOUR TEXT IN THIS CELL -->

In [7]:
# Write your code here
import numpy as np
from types import SimpleNamespace
import seaborn as sns
import matplotlib.pyplot as plt
import pandas as pd
from scipy.optimize import minimize
from scipy.optimize import fsolve
from problem1 import *
from problem1_3 import *

## 1. <a id='toc1_'></a>[Problem 1: Production economy and CO2 taxation](#toc0_)

Consider a production economy with two firms indexed by $j \in \{1,2\}$. Each produce its own good. They solve

$$
\begin{align*}
\max_{y_{j}}\pi_{j}&=p_{j}y_{j}-w_{j}\ell_{j}\\\text{s.t.}\;&y_{j}=A\ell_{j}^{\gamma}.
\end{align*}
$$

Optimal firm behavior is

$$
\begin{align*}
\ell_{j}^{\star}(w,p_{j})&=\left(\frac{p_{j}A\gamma}{w}\right)^{\frac{1}{1-\gamma}} \\
y_{j}^{\star}(w,p_{j})&=A\left(\ell_{j}^{\star}(w,p_{j})\right)^{\gamma}
\end{align*}
$$

The implied profits are

$$
\pi_{j}^*(w,p_{j})=\frac{1-\gamma}{\gamma}w\cdot\left(\frac{p_{j}A\gamma}{w}\right)^{\frac{1}{1-\gamma}}
$$

A single consumer supplies labor, and consumes the goods the firms produce. She also recieves the implied profits of the firm.<br>
She solves:

$$
\begin{align*}
U(p_1,p_2,w,\tau,T) = \max_{c_{1},c_{2},\ell} & \log(c_{1}^{\alpha}c_{2}^{1-\alpha})-\nu\frac{\ell^{1+\epsilon}}{1+\epsilon} \\
\text{s.t.}\,\,\,&p_{1}c_{1}+(p_{2}+\tau)c_{2}=w\ell+T+\pi_{1}^*(w,p_{1})+\pi_{2}^*(w,p_{2})
\end{align*}
$$

where $\tau$ is a tax and $T$ is lump-sum transfer. <br>
For a given $\ell$, it can be shown that optimal behavior is

$$
\begin{align*}
c_{1}(\ell)&=\alpha\frac{w\ell+T+\pi_{1}^*(w,p_{1})+\pi_{2}^*(w,p_{2})}{p_{1}} \\
c_{2}(\ell)&=(1-\alpha)\frac{w\ell+T+\pi_{1}^*(w,p_{1})+\pi_{2}^*(w,p_{2})}{p_{2}+\tau} \\
\end{align*}
$$
Such that optimal behavior is:
$$
\ell^* = \underset{\ell}{\arg\max} \log(\left(c_{1}(\ell)\right)^{\alpha}\cdot \left(c_{2}(\ell)\right)^{1-\alpha})-\nu\frac{\ell^{1+\epsilon}}{1+\epsilon} 
$$
With optimal consumption:
$$
\begin{align*}
c_1^*=c_{1}(\ell^*) \\
c_2^*=c_{2}(\ell^*)\\
\end{align*}
$$


The government chooses $\tau$ and balances its budget so $T=\tau c_2^*$. We initially set $\tau,T=0$.

Market clearing requires:

1. Labor market: $\ell^* = \ell_1^* + \ell_2^*$
1. Good market 1: $c_1^* = y_1^*$
1. Good market 2: $c_2^* = y_2^*$


**Question 1:** Check market clearing conditions for $p_1$ in `linspace(0.1,2.0,10)` and $p_2$ in `linspace(0.1,2.0,10)`. We choose $w=1$ as numeraire.

In [8]:
display(df_results)

p1        p2  labor_clearing  good1_clearing  good2_clearing
0   0.1  0.100000       -5.554690      -13.062278        0.100000
1   0.1  0.311111       -5.803897      -14.270006        0.311111
2   0.1  0.522222       -5.851920      -15.094386        0.522222
3   0.1  0.733333       -5.765498      -15.762883        0.733333
4   0.1  0.944444       -5.563581      -16.340265        0.944444
..  ...       ...             ...             ...             ...
95  2.0  1.155556        4.255356        0.877705        1.155556
96  2.0  1.366667        4.772789        0.854185        1.366667
97  2.0  1.577778        5.380573        0.832423        1.577778
98  2.0  1.788889        6.078467        0.812073        1.788889
99  2.0  2.000000        6.866303        0.792893        2.000000

[100 rows x 5 columns]

The above results show how the combinations of prices from the given linspace of (0.1, 2.0, 10), which creates an array of 10 equally spaced values between 0.1 and 2.0 for p1 and p2, affects market clearing. This allows the iteration over all posible combinations of p1 and p2 within the range and determines which combinations approimately satisfy the market clearing conditions. 

For example, row 0 displays p1 and p2 = 0.1. This combination of prices shows that labot demand exceeds labor supply by around 1.90. In the same row, production of the first good exceeds consumption of the first good by around 3.85 unit and production of the second good exceeds consumption by approximately 9.11 units. This means that the price combination of p1, p2 = 0.1 does not clear the market. 

None of the combinations of prices from the given linspace are found to clear the market. This only means that no combinations of prices in the linspace provide an equilibrium p1 and p2, but does not exclude the possibility of an equilibrium outside of the linspace. 

**Question 2:** Find the equilibrium prices $p_1$ and $p_2$.<br>
*Hint: you can use Walras' law to only check 2 of the market clearings*

In [12]:
from problem1_2 import *

# Initial guess for p1 and p2
initial_guess = [1.0, 1.0]

# Solve for market clearing prices
solution = fsolve(market_clearing_conditions, initial_guess, args=(1, par))

# Display the solution
print("Market clearing prices:")
print(f"p1 = {solution[0]}")
print(f"p2 = {solution[1]}")


Market clearing prices:
p1 = 0.7475486222511643
p2 = 0.7411786704525158


We solve the model and find the eqilibrium prices. These prices should accoring to Walras Law clear both the labor market and the market for both of the goods. We test whether the found equilibrum prices infact does clear the markets for labor and the two goods. 

In [ ]:
from problem1 import *
# Print results to check market clearing
print(f"Optimal labor (ell1*, ell2*): ({ell1_star}, {ell2_star})")
print(f"Optimal output (y1*, y2*): ({y1_star}, {y2_star})")
print(f"Implied profits (pi1*, pi2*): ({pi1_star}, {pi2_star})")
print(f"Total labor supplied (ell*): {ell_star_val}")
print(f"Optimal consumption (c1*, c2*): ({c1_star}, {c2_star})")


# Tolerance for approximate equality
tolerance = 1e-2

# Check market clearing conditions
labor_market_clears = np.abs(ell_star_val - (ell1_star + ell2_star)) < tolerance
goods_market_1_clears = np.abs(c1_star - y1_star) < tolerance
goods_market_2_clears = np.abs(c2_star - y2_star) < tolerance


print(f"Labor market clears: {labor_market_clears}")
print(f"Goods market 1 clears: {goods_market_1_clears}")
print(f"Goods market 2 clears: {goods_market_2_clears}")

The above output displays the optimal labor, output, profits, consumption, and total labour achieved with the equilibrium prices p1, p2, aswell as whether or not the labour market and the market for both goods clear given the equlibrium prices. Note that a tolerance level of 1e-2 (one hundreth) has been chosen.

We find that the equlibrium prices of p1 = 0.7475486222511643 and p2 = 0.7411786704525158 clear the labor market and the market for both good 1 and 2. 


Note we use a tolerance. We do this because we could not find a specfic result thus inserting a margin of error on 1e-2 (0.01) which resulted in market clearing. A method recommended by Co-pilot

***Pre text*** 

Assume the government care about the social welfare function:

$$
SWF = U - \kappa y_2^*
$$

Here $\kappa$ measures the social cost of carbon emitted by the production of $y_2$ in equilibrium.

**Question 3:** What values of $\tau$ and (implied) $T$ should the government choose to maximize $SWF$?

In [ ]:
# Print results
print(f"Optimal tau: {optimal_tau}")
print(f"Implied optimal T: {optimal_T}")
print(f"Optimal labor (ell*): {ell_star_val}")
print(f"Optimal consumption c1*: {c1_star}")
print(f"Optimal consumption c2*: {c2_star}")
print(f"Social Welfare Function (SWF): {SWF}")

We incorporate the definitions of kappa and the SWF into the model and solve for the optimal level of tau and thereafter the implied optimal level of T. The optimal level of tau is found to be 0.01950000000007 and the optimal implied T is found to be 0.013388916135360474. The optimal labor, consumption for c1 and c2, as well as the SWF is also displayed. 

The negative value for the SWF indicates the significance  of the social cost of carbon. The negative value indicates that even though optimal tax has been implemented, the welfare loss of carbon emmisions still reduce SWF substantially. 

## 2. <a id='toc2_'></a>[Problem 2: Career choice model](#toc0_)

Consider a graduate $i$ making a choice between entering $J$ different career tracks. <br>
Entering career $j$ yields utility $u^k_{ij}$. This value is unknown to the graduate ex ante, but will ex post be: <br>
$$
    u_{i,j}^k = v_{j} + \epsilon_{i,j}^k
$$

They know that $\epsilon^k_{i,j}\sim \mathcal{N}(0,\sigma^2)$, but they do not observe $\epsilon^k_{i,j}$ before making their career choice. <br>

Consider the concrete case of $J=3$ with:
$$
\begin{align*}
    v_{1} &= 1 \\
    v_{2} &= 2 \\
    v_{3} &= 3
\end{align*}
$$

If the graduates know the values of $v_j$ and the distribution of $\epsilon_{i,j}^k$, they can calculate the expected utility of each career track using simulation: <br>
$$
    \mathbb{E}\left[ u^k_{i,j}\vert v_j \right] \approx v_j + \frac{1}{K}\sum_{k=1}^K \epsilon_{i,j}^k
$$

In [ ]:
par = SimpleNamespace()
par.J = 3
par.N = 10
par.K = 10000

par.F = np.arange(1,par.N+1)
par.sigma = 2

par.v = np.array([1,2,3])
par.c = 1

**Question 1:** Simulate and calculate expected utility and the average realised utility for $K=10000$ draws, for each career choice $j$.


In [ ]:
from Eksamen2024 import question_2_1, question_2_2, question_2_3

# Question 2.1
expected_utilities, average_realized_utilities = question_2_1()
print("Expected Utilities for each career track:", expected_utilities)
print("Average Realized Utilities for each career track:", average_realized_utilities)

#### Conclusion
The overall trend shows that Career 3 is the most attractive option for graduates based on both expected and realized utilities. However, there are small discrepancies between expected and realized utilities for all career tracks, suggesting that while graduates' expectations are largely met, there are slight overestimations of the utility they would derive from each career track. This could be due to various factors such as initial enthusiasm, unforeseen challenges, or other external factors affecting their career experiences.

***Pre text***

**Now** consider a new scenario: Imagine that the graduate does not know $v_j$. The *only* prior information they have on the value of each job, comes from their $F_{i}$ friends that work in each career $j$. After talking with them, they know the average utility of their friends (which includes their friends' noise term), giving them the prior expecation: <br>
$$
\tilde{u}^k_{i,j}\left( F_{i}\right) = \frac{1}{F_{i}}\sum_{f=1}^{F_{i}} \left(v_{j} + \epsilon^k_{f,j}\right), \; \epsilon^k_{f,j}\sim \mathcal{N}(0,\sigma^2)
$$
For ease of notation consider that each graduate have $F_{i}=i$ friends in each career. <br>

For $K$ times do the following: <br>
1. For each person $i$ draw $J\cdot F_i$ values of $\epsilon_{f,j}^{k}$, and calculate the prior expected utility of each career track, $\tilde{u}^k_{i,j}\left( F_{i}\right)$. <br>
Also draw their own $J$ noise terms, $\epsilon_{i,j}^k$
1. Each person $i$ chooses the career track with the highest expected utility: $$j_i^{k*}= \arg\max_{j\in{1,2\dots,J}}\left\{ \tilde{u}^k_{i,j}\left( F_{i}\right)\right\} $$
1. Store the chosen careers: $j_i^{k*}$, the prior expectation of the value of their chosen career: $\tilde{u}^k_{i,j=j_i^{k*}}\left( F_{i}\right)$, and the realized value of their chosen career track: $u^k_{i,j=j_i^{k*}}=v_{j=j_i^{k*}}+\epsilon_{i,j=j_i^{k*}}^k$.

Chosen values will be: <br>
$i\in\left\{1,2\dots,N\right\}, N=10$ <br>
$F_i = i$<br>
So there are 10 graduates. The first has 1 friend in each career, the second has 2 friends, ... the tenth has 10 friends.

**Question 2:** Simulate and visualize: For each type of graduate, $i$, the share of graduates choosing each career, the average subjective expected utility of the graduates, and the average ex post realized utility given their choice. <br>
That is, calculate and visualize: <br>
$$
\begin{align*}
    \frac{1}{K} \sum_{k=1}^{K} \mathbb{I}\left\{ j=j_i^{k*} \right\}  \;\forall j\in\left\{1,2,\dots,J\right\}
\end{align*}
$$
$$
\begin{align*}
    \frac{1}{K} \sum_{k=1}^{K} \tilde{u}^k_{ij=j_i^{k*}}\left( F_{i}\right)
\end{align*}
$$
And 
$$
\begin{align*}
    \frac{1}{K} \sum_{k=1}^{K} u^k_{ij=j_i^{k*}} 
\end{align*}
$$
For each graduate $i$.

In [ ]:
# Question 2.2
choices, prior_expected_utilities, realized_utilities = question_2_2()
career_counts = np.bincount(choices, minlength=3)
average_prior_expected_utilities = np.mean(prior_expected_utilities)
average_realized_utilities = np.mean(realized_utilities)

print("Career Counts:", career_counts)
print("Average Prior Expected Utilities:", average_prior_expected_utilities)
print("Average Realized Utilities:", average_realized_utilities)

# Prepare data for visualization
df_share = pd.DataFrame({
    'Career': [f'Career {i+1}' for i in range(3)],
    'Share': career_counts / 10
})

df_utility = pd.DataFrame({
    'Type': ['Average Prior Expected Utility', 'Average Realized Utility'],
    'Utility': [average_prior_expected_utilities, average_realized_utilities]
})

# Set Seaborn style
sns.set(style="whitegrid")

# Create the plots
fig, axes = plt.subplots(1, 2, figsize=(14, 6))

# Plot the share of graduates choosing each career
sns.barplot(x='Career', y='Share', data=df_share, palette="viridis", ax=axes[0])
axes[0].set_xlabel('Career Tracks')
axes[0].set_ylabel('Share of Graduates Choosing Each Career')
axes[0].set_title('Share of Graduates Choosing Each Career Track')

# Plot the average subjective expected utility
sns.barplot(x='Type', y='Utility', data=df_utility, palette="magma", ax=axes[1])
axes[1].set_ylabel('Utility')
axes[1].set_title('Average Utilities')

plt.tight_layout()
plt.show()

#### Conclusion

Career 3 is the most preferred career track among graduates, followed by Career 1, with Career 2 being the least preferred. The Graduates tend to overestimate the utility they expect to derive from their chosen career tracks. The average prior expected utility is higher than the average realized utility, indicating that the actual experiences of graduates are not as favorable as they had anticipated. The significant discrepancy between expected and realized utilities suggests that graduates might need better information or guidance when choosing career tracks to align their expectations more closely with reality.

***Pre text***

After a year of working in their career, the graduates learn $u^k_{ij}$ for their chosen job $j_i^{k*}$ perfectly. <br>
The can switch to one of the two remaining careers, for which they have the same prior as before, but it will now include a switching cost of $c$ which is known.
Their new priors can be written as: 
$$
\tilde{u}^{k,2}_{ij}\left( F_{i}\right) = \begin{cases}
            \tilde{u}^k_{ij}\left( F_{i}\right)-c & \text{if } j \neq j_i^{k*} \\
            u_{ij=j_i^{k*}} & \text{if } j = j_i^{k*}
        \end{cases}
$$

We will set $c=1$.

Their realized utility will be: <br>
$$
u^{k,2}_{ij}= \begin{cases}
            u_{ij}^k -c & \text{if } j \neq j_i^{k*} \\
            u_{ij=j_i^{k*}} & \text{if } j = j_i^{k*}
        \end{cases}
$$

**Question 3:** Following the same approach as in question 2, find the new optimal career choice for each $i$, $k$. Then for each $i$, calculate the average subjective expected utility from their new optimal career choice, and the ex post realized utility of that career. Also, for each $i$, calculate the share of graduates that chooses to switch careers, conditional on which career they chose in the first year. <br>

In [ ]:
# Question 2.3
new_choices, new_expected_utilities, new_realized_utilities, switch_decisions = question_2_3()
career_counts = np.bincount(new_choices, minlength=3)
average_new_expected_utilities = np.mean(new_expected_utilities)
average_new_realized_utilities = np.mean(new_realized_utilities)
switch_share = np.mean(switch_decisions)

print("New Career Counts:", career_counts)
print("Average New Expected Utilities:", average_new_expected_utilities)
print("Average New Realized Utilities:", average_new_realized_utilities)
print("Share of Graduates Switching Careers:", switch_share)

# Prepare data for visualization
df_share = pd.DataFrame({
    'Career': [f'Career {i+1}' for i in range(3)],
    'Share': career_counts / 10
})

df_utility = pd.DataFrame({
    'Type': ['Average New Expected Utility', 'Average New Realized Utility'],
    'Utility': [average_new_expected_utilities, average_new_realized_utilities]
})

# Set Seaborn style
sns.set(style="whitegrid")

# Create the plots
fig, axes = plt.subplots(1, 2, figsize=(14, 6))

# Plot the share of graduates choosing each career after switching
sns.barplot(x='Career', y='Share', data=df_share, palette="viridis", ax=axes[0])
axes[0].set_xlabel('Career Tracks')
axes[0].set_ylabel('Share of Graduates Choosing Each Career')
axes[0].set_title('Share of Graduates Choosing Each Career Track After Switching')

# Plot the average subjective expected utility after switching
sns.barplot(x='Type', y='Utility', data=df_utility, palette="magma", ax=axes[1])
axes[1].set_ylabel('Utility')
axes[1].set_title('Average Utilities After Switching')

plt.tight_layout()
plt.show()

#### Conclusion
Career 3 remains the most preferred career track even after considering switching costs, followed by Career 2, with Career 1 being the least preferred. Both the new expected and realized utilities are higher than the initial ones, suggesting that the switching led to better outcomes for the graduates. The close alignment between new expected and realized utilities indicates that graduates' new decisions were well-informed and realistic. The fact that 50% of the graduates switched careers highlights the importance of reassessment and the positive impact of having the flexibility to switch tracks.

## 3. <a id='toc3_'></a>[Problem 3: Barycentric interpolation](#toc0_)

**Problem:** We have a set of random points in the unit square,

$$
\mathcal{X} = \{(x_1,x_2)\,|\,x_1\sim\mathcal{U}(0,1),x_2\sim\mathcal{U}(0,1)\}.
$$

For these points, we know the value of some function $f(x_1,x_2)$,

$$
\mathcal{F} = \{f(x_1,x_2) \,|\, (x_1,x_2) \in \mathcal{X}\}.
$$

Now we want to approximate the value $f(y_1,y_2)$ for some  $y=(y_1,y_2)$, where $y_1\sim\mathcal{U}(0,1)$ and $y_2\sim\mathcal{U}(0,1)$.

**Building block I**

For an arbitrary triangle $ABC$ and a point $y$, define the so-called barycentric coordinates as:

$$
\begin{align*}
  r^{ABC}_1 &= \frac{(B_2-C_2)(y_1-C_1) + (C_1-B_1)(y_2-C_2)}{(B_2-C_2)(A_1-C_1) + (C_1-B_1)(A_2-C_2)} \\
  r^{ABC}_2 &= \frac{(C_2-A_2)(y_1-C_1) + (A_1-C_1)(y_2-C_2)}{(B_2-C_2)(A_1-C_1) + (C_1-B_1)(A_2-C_2)} \\
  r^{ABC}_3 &= 1 - r_1 - r_2.
\end{align*}
$$

If $r^{ABC}_1 \in [0,1]$, $r^{ABC}_2 \in [0,1]$, and $r^{ABC}_3 \in [0,1]$, then the point is inside the triangle.

We always have $y = r^{ABC}_1 A + r^{ABC}_2 B + r^{ABC}_3 C$.

**Building block II**

Define the following points:

$$
\begin{align*}
A&=\arg\min_{(x_{1},x_{2})\in\mathcal{X}}\sqrt{\left(x_{1}-y_{1}\right)^{2}+\left(x_{2}-y_{2}\right)^{2}}\text{ s.t. }x_{1}>y_{1}\text{ and }x_{2}>y_{2}\\
B&=\arg\min_{(x_{1},x_{2})\in\mathcal{X}}\sqrt{\left(x_{1}-y_{1}\right)^{2}+\left(x_{2}-y_{2}\right)^{2}}\text{ s.t. }x_{1}>y_{1}\text{ and }x_{2}<y_{2}\\
C&=\arg\min_{(x_{1},x_{2})\in\mathcal{X}}\sqrt{\left(x_{1}-y_{1}\right)^{2}+\left(x_{2}-y_{2}\right)^{2}}\text{ s.t. }x_{1}<y_{1}\text{ and }x_{2}<y_{2}\\
D&=\arg\min_{(x_{1},x_{2})\in\mathcal{X}}\sqrt{\left(x_{1}-y_{1}\right)^{2}+\left(x_{2}-y_{2}\right)^{2}}\text{ s.t. }x_{1}<y_{1}\text{ and }x_{2}>y_{2}.
\end{align*}
$$

**Algorithm:**

1. Compute $A$, $B$, $C$, and $D$. If not possible return `NaN`.
1. If $y$ is inside the triangle $ABC$ return $r^{ABC}_1 f(A) + r^{ABC}_2 f(B) + r^{ABC}_3 f(C)$.
1. If $y$ is inside the triangle $CDA$ return $r^{CDA}_1 f(C) + r^{CDA}_2 f(D) + r^{CDA}_3 f(A)$.
1. Return `NaN`.



**Sample:**

In [ ]:
rng = np.random.default_rng(2024)

X = rng.uniform(size=(50,2))
y = rng.uniform(size=(2,))

**Questions 1:** Find $A$, $B$, $C$ and $D$. Illustrate these together with $X$, $y$ and the triangles $ABC$ and $CDA$.

In [ ]:
from problem3 import *

# Find the closest points
A, B, C, D = find_closest_points(X, y)

# Set seaborn style
sns.set(style="whitegrid")

# Plot the points
plt.figure(figsize=(9, 7))

# Plot points in X
sns.scatterplot(x=X[:, 0], y=X[:, 1], color='blue', label='Points in X', s=100)

# Plot the point y and points A, B, C, D
sns.scatterplot(x=[y[0]], y=[y[1]], color='red', label='Point y', s=200, zorder=5)
sns.scatterplot(x=[A[0]], y=[A[1]], color='green', label='Point A (Quadrant 1)', s=200, zorder=5)
sns.scatterplot(x=[B[0]], y=[B[1]], color='purple', label='Point B (Quadrant 2)', s=200, zorder=5)
sns.scatterplot(x=[C[0]], y=[C[1]], color='orange', label='Point C (Quadrant 3)', s=200, zorder=5)
sns.scatterplot(x=[D[0]], y=[D[1]], color='brown', label='Point D (Quadrant 4)', s=200, zorder=5)

# Draw the triangles
triangle_ABC = plt.Polygon([A, B, C], color='green', alpha=0.2, label='Triangle ABC')
triangle_CDA = plt.Polygon([C, D, A], color='brown', alpha=0.2, label='Triangle CDA')
plt.gca().add_patch(triangle_ABC)
plt.gca().add_patch(triangle_CDA)

plt.xlabel('x1')
plt.ylabel('x2')
plt.title('Points and Triangles')
plt.legend(bbox_to_anchor=(1.05, 1), loc='upper left')
plt.grid(True)
plt.show()

**Question 2:** Compute the barycentric coordinates of the point $y$ with respect to the triangles $ABC$ and $CDA$. Which triangle is $y$ located inside?

In [ ]:
from problem3 import *
# Print results
print("Barycentric coordinates of y with respect to ABC:")
print(f"lambda1: {lambda1_ABC}, lambda2: {lambda2_ABC}, lambda3: {lambda3_ABC}")
print(f"Is y inside ABC? {'Yes' if is_inside_ABC else 'No'}")

print("\nBarycentric coordinates of y with respect to CDA:")
print(f"lambda1: {lambda1_CDA}, lambda2: {lambda2_CDA}, lambda3: {lambda3_CDA}")
print(f"Is y inside CDA? {'Yes' if is_inside_CDA else 'No'}")

Now consider the function:
$$
f(x_1,x_2) = x_1 \cdot x_2
$$

In [ ]:
f = lambda x: x[0]*x[1]
F = np.array([f(x) for x in X])

**Question 3:** Compute the approximation of $f(y)$ using the full algorithm. Compare with the true value.

In [ ]:
from problem3 import *

# Compute barycentric coordinates for y with respect to ABC and CDA
lambda1_ABC, lambda2_ABC, lambda3_ABC = barycentric_coordinates(y, A, B, C)
lambda1_CDA, lambda2_CDA, lambda3_CDA = barycentric_coordinates(y, C, D, A)

# Compute f values at A, B, C, D
f_A = f(A)
f_B = f(B)
f_C = f(C)
f_D = f(D)

# Interpolate f(y) using ABC
f_y_ABC = lambda1_ABC * f_A + lambda2_ABC * f_B + lambda3_ABC * f_C

# Interpolate f(y) using CDA
f_y_CDA = lambda1_CDA * f_C + lambda2_CDA * f_D + lambda3_CDA * f_A

# Determine which triangle y is inside based on barycentric coordinates
if (0 <= lambda1_ABC <= 1) and (0 <= lambda2_ABC <= 1) and (0 <= lambda3_ABC <= 1):
    f_y_approx = f_y_ABC
elif (0 <= lambda1_CDA <= 1) and (0 <= lambda2_CDA <= 1) and (0 <= lambda3_CDA <= 1):
    f_y_approx = f_y_CDA
else:
    f_y_approx = None

# Compute the true value of f(y)
f_y_true = f(y)

# Print results
print("Barycentric coordinates of y with respect to ABC:")
print(f"lambda1: {lambda1_ABC}, lambda2: {lambda2_ABC}, lambda3: {lambda3_ABC}")

print("\nBarycentric coordinates of y with respect to CDA:")
print(f"lambda1: {lambda1_CDA}, lambda2: {lambda2_CDA}, lambda3: {lambda3_CDA}")

print("\nInterpolated value of f(y):")
print(f"f_y_approx: {f_y_approx}")

print("\nTrue value of f(y):")
print(f"f_y_true: {f_y_true}")

# Compare the interpolated value with the true value
if f_y_approx is not None:
    print("\nComparison:")
    print(f"Error: {abs(f_y_true - f_y_approx)}")
else:
    print("Point y is not inside any of the triangles ABC or CDA.")

#### Conclusion
The approximation of \( f(y) \) using barycentric coordinates is effective, providing a value close to the true value with a small error margin. The small difference between the true value and the approximated value indicates that the approximation method is fairly accurate, though there is a slight overestimation in this case. The barycentric coordinate method proves to be a useful tool for approximating function values at given points within triangles, with practical accuracy for many applications.


**Question 4:** Repeat question 3 for all points in the set $Y$.

In [ ]:
Y = [(0.2,0.2),(0.8,0.2),(0.8,0.8),(0.8,0.2),(0.5,0.5)]

In [ ]:
from problem3_2 import *

# Display the results
for result in results:
    print(f"Point y: {result['y']}")
    print(f"Interpolated value of f(y): {result['f_y_approx']}")
    print(f"True value of f(y): {result['f_y_true']}")
    print(f"Error: {result['error']}")
    print(f"Inside ABC: {result['inside_ABC']}")
    print(f"Inside CDA: {result['inside_CDA']}")
    print("\n" + "-"*40 + "\n")

# Set seaborn style
sns.set(style="whitegrid")

# Plot the points and triangles
plt.figure(figsize=(9, 7))

# Plot points in X
sns.scatterplot(x=X[:, 0], y=X[:, 1], color='blue', label='Points in X', s=100)

# Plot the point y and set Y
for result in results:
    if result['f_y_approx'] is not None:
        label = f"Point {result['y']}, Error: {result['error']:.2e}"
    else:
        label = f"Point {result['y']}, Outside triangles"
    sns.scatterplot(x=[result['y'][0]], y=[result['y'][1]], color='red', label=label, s=200)

# Plot points A, B, C, D
sns.scatterplot(x=[A[0]], y=[A[1]], color='green', label='Point A (Quadrant 1)', s=200)
sns.scatterplot(x=[B[0]], y=[B[1]], color='purple', label='Point B (Quadrant 2)', s=200)
sns.scatterplot(x=[C[0]], y=[C[1]], color='orange', label='Point C (Quadrant 3)', s=200)
sns.scatterplot(x=[D[0]], y=[D[1]], color='brown', label='Point D (Quadrant 4)', s=200)

# Draw the triangles
triangle_ABC = plt.Polygon([A, B, C], color='green', alpha=0.2, label='Triangle ABC')
triangle_CDA = plt.Polygon([C, D, A], color='brown', alpha=0.2, label='Triangle CDA')
plt.gca().add_patch(triangle_ABC)
plt.gca().add_patch(triangle_CDA)

plt.xlabel('x1')
plt.ylabel('x2')
plt.title('Points and Triangles')
plt.legend(bbox_to_anchor=(1.05, 1), loc='upper left')
plt.grid(True)
plt.show()

#### Conclusion

- **Point 1**: The approximation is highly accurate, with the approximated value being very close to the true value.
- **Point 2**: There is a significant overestimation in the approximated value compared to the true value.
- **Point 3**: There is a noticeable underestimation in the approximated value compared to the true value.
- **Overall Accuracy**: The MSE indicates that while the approximation method is effective, it has a moderate error level, with some points being overestimated and others underestimated.
